cloud-init.pl

```diff
@@ -130,18 +145,16 @@ sub cloud_init {
     my $pubkeys = get_data($host, 'meta-data/public-keys');
     chomp($pubkeys);
     install_pubkeys $pubkeys;
-    set_hostname get_default_fqdn;
-
-    if (defined($data)) {
-        if ($data =~ /^#cloud-config/) {
-            $data = CPAN::Meta::YAML->read_string($data)->[0];
-            apply_user_data $data;
-        } elsif ($data =~ /^#\!/) {
-            my ($fh, $filename) = tempfile("/tmp/cloud-config-XXXXXX");
-            print $fh $data;
-            chmod(0700, $fh);
-            close $fh;
-            system("sh -c \"$filename && rm $filename\"");
+
+    if ($data =~ /^#cloud-config/) {
+        $data = CPAN::Meta::YAML->read_string($data)->[0];
+        apply_user_data($data);
+    } else {
+        set_hostname(get_default_fqdn);
+        add_etc_hosts_entry(get_default_fqdn);
+
+        if ($data =~ /^#\!/) {
```

Comment:
I wonder if the structure here shouldn't be 

```perl
if ($data =~ /^#cloud-config/) {
   $data = CPAN::Meta::YAML->read_string($data)->[0];
   apply_user_data($data);
} else if ($data =~ /^#\!/) {
  run_user_script($data);
} else {
  # If no user data is provided, assume default behavior of setting the hostname and
  # creating an /etc/hosts entry
  set_hostname(get_default_fqdn);
  add_etc_hosts_entry(get_default_fqdn);
}
```

Reply:
> I got sidetracked and forgot about this one, sorry.

No worries, this is not exactly a critical issue. I just thought I'd ask if you had gotten around to taking a look at this pull request.

Reply:
Ah, I see what you mean. At the moment, `set_hostname` is called unconditionally:
[cloud-init.pl#L133](https://github.com/exoscale/openbsd-cloud-init/blob/fc50a98dd79f3c2bdf20ecd9608e02d8798dcc8f/cloud-init.pl#L133)

However, `/etc/hosts` is only updated in `apply_user_data`:
[https://github.com/exoscale/openbsd-cloud-init/blob/fc50a98dd79f3c2bdf20ecd9608e02d8798dcc8f/cloud-init.pl#L70-L85](https://github.com/exoscale/openbsd-cloud-init/blob/fc50a98dd79f3c2bdf20ecd9608e02d8798dcc8f/cloud-init.pl#L70-L85)

I think calling `set_hostname` unconditionally (the way it is done now) is correct.

Regarding `add_etc_hosts_entry`, I am not quite as sure. I think my current suggestion is easier to reason about as a user than if we were to disable these calls if the user supplies a user script.

The best way of handling this would be to see how cloud-init handles user scripts in general. I'll investigate later.

Reply:
Thanks for checking, I'll merge as it is now.